File: local/src/lib/awormap.js
// Optimized Add-Wins Observed-Remove Map

import { DotContext } from "./dotcontext";
import { Product } from "./product";

/**
 * Add-Wins Observed-Remove Map (AWORMap) - A map that stores [element_id, dot_value] =>
 * element and keeps causal context using a DotContext that maps element_id => dot_value.
 * This is optimized thanks to the DotContext implementation.
 */
export class AWORMap {
    #id; // this is most likely deprecated
    /**
     * Map that stores [element_id, dot_value] => element
     */
    #map = new Map();
    /**
     * DotContext that stores element_id => dot_value
     */
    #cc = new DotContext();

    constructor(id) {
        this.#id = id;
    }

    /**
     *
     * @returns {Map} map
     */
    getMap() {
        const res = this.#map;
        return res;
    }

    /**
     *
     * @returns {DotContext} causal context
     */
    getCC() {
        const res = this.#cc;
        return res;
    }

    elements() {
        return this.#map.keys();
    }

    /**
     * Returns the elements of the map that are in the causal context
     *
     * @returns {IterableIterator} elements
     */
    values() {
        const cc = this.#cc.getCC();
        return new Map(
            [...this.#map].filter(([key, value]) => cc.get(key[0]) == key[1]),
        ).values();
    }

    /**
     * Returns a map of the elements that are in the causal context
     *
     * @returns {Map} elements
     */
    flatten() {
        const cc = this.#cc.getCC();
        return new Map(
            [...this.#map].filter(([key, value]) => cc.get(key[0]) == key[1]),
        );
    }

    /**
     * Adds a new element to the map, keeping the causal context
     *
     * @param {*} element_id Identifier of the element
     * @param {*} element Element to be added
     */
    add(element_id, element) {
        const dot = this.#cc.makeDot(element_id);
        this.#map.set([dot[0], dot[1]], element);
    }

    /**
     * Removes an element from the map, keeping the causal context
     *
     * @param {*} element_id Identifier of the element
     */
    rm(element_id) {
        for (const [key, value] of this.#map) {
            if (key[0] == element_id) this.#map.delete(key);
        }
    }

    get(element_id) {
        for (const [key, value] of this.flatten()) {
            if (key[0] == element_id) return value;
        }
        return null;
    }

    /**
     * Join two AWORMaps
     *
     * @param {AWORMap} awormap
     */
    join(awormap) {
        this.#map = new Map(
            [...this.#map].filter(
                ([key, value]) =>
                    awormap.getMap().has(key) || !awormap.getCC().dotIn(key),
            ),
        );

        for (const [key, value] of awormap.getMap()) {
            if (!this.#cc.dotIn(key[0], key[1])) {
                this.#map.set(key, value);
            }
        }

        this.#cc.join(awormap.getCC());
    }

    fromJSON(id, products) {
        this.#id = id;

        for (const jproduct of products.map) {
            const dot = [jproduct.name, jproduct.context];
            const product = new Product();

            product.fromJSON(jproduct);

            this.#map.set(dot, product);
        }

        // FIXME: this seems to be an edge case, where due to the fact that we are reconstructing a map from a JSON, we need to
        // insert the dots in the CC, but using insertDot() won't work due to how the compacting works, there may be a better way
        // to do this, or not it could be fine as it is as it doesn't seem to be a common issue
        for (const dot of products.context.cc)
            this.#cc.getCC().set(dot[0], dot[1]);

        if (products.context.dc.length > 0)
            for (const dot of products.context.dc)
                this.#cc.insertDot(dot, false);

    }

    toJSON() {
        const res = {
            products: {
                map: [],
                context: {
                },
            },
        };

        for (const [key, value] of this.#map) {
            const product = value.toJSON();
            product.name = key[0];
            product.context = key[1];
            res.products.map.push(product);
        }

        res.products.context = this.#cc.toJSON()
        
        return res;
    }
}

File: cloud/src/main/java/server/model/AWORMap.java
package server.model;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

public class AWORMap {
    private String id; // This is most likely deprecated
    /**
     * Map that stores [element_id, dot_value] => element
     */
    private Map<int[], Product> map = new HashMap<>();
    /**
     * DotContext that stores element_id => dot_value
     */
    private DotContext cc;

    /**
     * Constructor with an identifier
     *
     * @param id Identifier
     */
    public AWORMap(String id) {
        this.id = id;
    }

    /**
     * Constructor with products
     *
     * @param products Map of products
     */
    public AWORMap(Map<int[], Product> products) {
        for (Map.Entry<int[], Product> entry : products.entrySet()) {
            int[] key = entry.getKey();
            Product value = entry.getValue();
            this.add(key[0], value);
        }

        this.cc = new DotContext(products.context);
    }

    /**
     * Get the map
     *
     * @return Map
     */
    public Map<int[], Product> getMap() {
        return map;
    }

    /**
     * Get the causal context
     *
     * @return Causal context
     */
    public DotContext getCC() {
        return cc;
    }

    /**
     * Get the elements of the map
     *
     * @return Iterator of elements
     */
    public Iterator<int[]> elements() {
        return map.keySet().iterator();
    }

    /**
     * Get the values of the map that are in the causal context
     *
     * @return Iterator of values
     */
    public Iterator<Product> values() {
        Map<int[], Product> filteredMap = new HashMap<>();
        for (Map.Entry<int[], Product> entry : map.entrySet()) {
            int[] key = entry.getKey();
            Product value = entry.getValue();
            if (cc.getCC().containsKey(key[0]) && cc.getCC().get(key[0]) == key[1]) {
                filteredMap.put(key, value);
            }
        }
        return filteredMap.values().iterator();
    }

    /**
     * Add a new element to the map, keeping the causal context
     *
     * @param element_id Identifier of the element
     * @param element    Element to be added
     */
    public void add(int element_id, Product element) {
        int[] dot = cc.makeDot(element_id);
        map.put(dot, element);
    }

    /**
     * Remove an element from the map, keeping the causal context
     *
     * @param element_id Identifier of the element
     */
    public void rm(int element_id) {
        Iterator<Map.Entry<int[], Product>> iterator = map.entrySet().iterator();
        while (iterator.hasNext()) {
            Map.Entry<int[], Product> entry = iterator.next();
            int[] key = entry.getKey();
            if (key[0] == element_id) {
                iterator.remove();
            }
        }
    }

    /**
     * Get an element from the map
     *
     * @param element_id Identifier of the element
     * @return Element
     */
    public Product get(int element_id) {
        for (Map.Entry<int[], Product> entry : map.entrySet()) {
            int[] key = entry.getKey();
            if (key[0] == element_id) {
                return entry.getValue();
            }
        }
        return null;
    }

    /**
     * Join two AWORMaps
     *
     * @param awormap AWORMap to join
     */
    public void join(AWORMap awormap) {
        Iterator<Map.Entry<int[], Product>> iterator = map.entrySet().iterator();
        while (iterator.hasNext()) {
            Map.Entry<int[], Product> entry = iterator.next();
            int[] key = entry.getKey();
            Product value = entry.getValue();
            if (!awormap.getMap().containsKey(key) || !awormap.getCC().dotIn(key)) {
                iterator.remove();
            }
        }

        for (Map.Entry<int[], Product> entry : awormap.getMap().entrySet()) {
            int[] key = entry.getKey();
            Product value = entry.getValue();
            if (!cc.dotIn(key, value.getCounter())) {
                map.put(key, value);
            }
        }

        cc.join(awormap.getCC());
    }
}

